software/controller/lib/hal/spi_chain.tpp

```diff
@@ -17,46 +17,91 @@ limitations under the License.
+  ResetSlaves(reset_command, response_buffer, sizeof(response_buffer));
+
+  ParseProbeResponse(response_buffer, sizeof(response_buffer));
+}
+
+template<size_t MaxSlaves, size_t MaxRequestsPerSlave>
+void DaisyChain<MaxSlaves, MaxRequestsPerSlave>::SetNullCommand(uint8_t null_command){
+  null_command_ = null_command;
+};
+
+template<size_t MaxSlaves, size_t MaxRequestsPerSlave>
+void DaisyChain<MaxSlaves, MaxRequestsPerSlave>::FlushSlavesData(){
+  // Send a bunch of null commands in order to flush all of the data in the daisy chain and any data
+  // currently being returned by the slaves (in case the controller was halted in a weird state).
+  uint8_t flush_buffer[MaxSlaves*5] = {0};
+  for(size_t i = 0 ; i < sizeof(flush_buffer) ; i++){
+    flush_buffer[i] = null_command_;
+  }
+  uint8_t response_buffer[sizeof(flush_buffer)] = {0};
+  SendDataWithBusyWait(flush_buffer, response_buffer, sizeof(flush_buffer));
+};
+
+//
+template<size_t MaxSlaves, size_t MaxRequestsPerSlave>
+void DaisyChain<MaxSlaves, MaxRequestsPerSlave>::ResetSlaves(uint8_t reset_command, uint8_t *response_buffer, size_t length){
+  uint8_t reset_buffer[length] = {0};
+  for(size_t i = 0 ; i < sizeof(reset_buffer) ; i++){
+    reset_buffer[i] = reset_command;
+  }
+  SendDataWithBusyWait(reset_buffer, response_buffer, length);
+};
+
+template<size_t MaxSlaves, size_t MaxRequestsPerSlave>
+void DaisyChain<MaxSlaves, MaxRequestsPerSlave>::ParseProbeResponse(uint8_t *response_buffer, size_t length) {
+  // At this point, the response buffer should be filled with as many null commands as there are
+  // slaves. Parse it to determine how many slaves there are.
+  num_slaves_ = 0;
+  for(size_t i = 0 ; i < length ; i++){
+    if (response_buffer[i] == null_command_){
+      num_slaves_++;
+    }
+  }
+
+  // If the response buffer is full of null commands, this most likely means the
+  // daisy chain is broken (more slaves than we expect)
+  if(num_slaves_==sizeof(response_buffer)) {
+    num_slaves_ = 0;
+  }
 }
 
-template<size_t MaxSlaves>
-bool DaisyChain<MaxSlaves>::SendRequest(const Request &request, size_t slave) {
+template<size_t MaxSlaves, size_t MaxRequestsPerSlave>
+bool DaisyChain<MaxSlaves, MaxRequestsPerSlave>::SendRequest(const Request &request, size_t slave) {
   // Ensure thread safety as this function might be called from a timer interrupt
   // as well as the main loop.
   BlockInterrupts block;
   // Check that everything is in order to proceed: the slave is valid, there is room in the
   // request queue for that slave and in the command buffer, and the provided data looks OK
-  if(slave >= num_slaves_ || queue_count_[slave] >= MaxQueueLength ||
+  if(slave >= num_slaves_ || queue_count_[slave] >= MaxRequestsPerSlave ||
        request.length > CommandBufferSize - command_buffer_count_ ||
        request.command == nullptr || request.length == 0){
     /// \TODO log an error (along with its nature?)
     queueing_errors_.set(queueing_errors_.get() + 1);
```

Comment:
Another way to do this would be to use a Debug 32 bit primitive and link it by reference to a regular old `uint32` you have as member. Then you could just do `errors_++` here.